gcc/d/d-codegen.cc

```diff
@@ -2911,26 +3134,165 @@ build_float_modulus (tree type, tree arg0, tree arg1)
 tree
 build_typeinfo (Type *t)
 {
-  tree tinfo = t->getInternalTypeInfo (NULL)->toElem (current_irstate);
+  tree tinfo = t->getInternalTypeInfo (NULL)->toElem(NULL);
   gcc_assert (POINTER_TYPE_P (TREE_TYPE (tinfo)));
   return tinfo;
 }
 
-// Build LABEL_DECL at location LOC for IDENT given.
+// Check that a new jump at FROM to a label at TO is OK.
+
+void
+check_goto(Statement *from, Statement *to)
+{
+  d_label_entry *ent = cfun->language->labels->get(to);
+  gcc_assert(ent != NULL);
+
+  // If the label hasn't been defined yet, defer checking.
+  if (! DECL_INITIAL (ent->label))
+    {
+      d_label_use_entry *fwdref = ggc_alloc<d_label_use_entry>();
+      fwdref->level = current_binding_level;
+      fwdref->statement = from;
+      fwdref->next = ent->fwdrefs;
+      ent->fwdrefs = fwdref;
+      return;
+    }
+
+  if (ent->in_try_scope)
+    from->error("cannot goto into try block");
+  else if (ent->in_catch_scope)
+    from->error("cannot goto into catch block");
+}
+
+// Check that a previously seen jumps to a newly defined label is OK.
+
+static void
+check_previous_goto(Statement *s, d_label_use_entry *fwdref)
+{
+  bool complain = false;
+
+  for (binding_level *b = current_binding_level; b ; b = b->level_chain)
+    {
+      if (b == fwdref->level)
+	break;
+
+      if ((b->kind == level_try || b->kind == level_catch) && !complain)
+	{
+	  if (s->isLabelStatement())
+	    {
+	      if (b->kind == level_try)
+		fwdref->statement->error("cannot goto into try block");
+	      else
+		fwdref->statement->error("cannot goto into catch block");
+	    }
+	  else if (s->isCaseStatement())
+	    s->error("case cannot be in different try block level from switch");
+	  else if (s->isDefaultStatement())
+	    s->error("default cannot be in different try block level from switch");
+	  else
+	    gcc_unreachable();
+
+	  complain = true;
```

Comment:
Artifact of some other way to test what kind of label we have.  Having another look, I don't think limiting the errors to just the first is any helpful either.